### DISCUSSION TITLE
Use pre-commit manager with the git-refs datasource

### DISCUSSION THREAD

**Discussion by eyalzek** [renovatebot/renovate#10846] [Request Help]
**What would you like Renovate to be able to do?**
we would like to use pre-commit with ["bleeding edge versioning"](https://pre-commit.com/#pre-commit-autoupdate), so basically the latest git ref of the pre-commit repo. At the moment the `.pre-commit-config.yaml` file is being matched by the `pre-commit` manager which uses the `github-tags` - although this is the recommended way to go according to pre-commit, we would like to avoid tagging. 

Is it actually possible at the moment to override the datasource used by a manager?

**Comment by viceice** [comment]
Sure, use a package rule, match your deps / manager and assign a new datasource.

**Reply by eyalzek** [reply]
But how do you assign a new datasource?

```json
"packageRules": [
        {
          "matchManagers": ["pre-commit"],
          "datasource": "git-refs"
        }
]
```

doesn't seem to work and I can't find a relevant configuration option

**Reply by eyalzek** [reply]
Oh actually it does work, but the log output still shows `github-tags`:

```json
"pre-commit": [
  {
    "packageFile": ".pre-commit-config.yaml",
    "deps": [
      {
        "datasource": "github-tags",
        ...
        "versioning": "git"
      }
    ]
  }
]
```

**Reply by viceice** [reply]
Yes, the output is before applying any rules. It's just the extract phase. 🤷‍♂️

**Reply by raspla** [reply]
This does not seem to work with `gitlabci-include` manager: `"validationMessage": "Invalid configuration option: packageRules[0].datasource"`

```json
  "packageRules": [
    {
      "matchManagers": ["gitlabci-include"],
      "datasource": "git-refs"
    }
  ]
```

**Comment by eyalzek** [comment]
For posterity, considering you have pre-commit file with the following format:

```yaml
---
repos:
  - repo: git@github.com:foo/bar.git
    rev: xyz
    hooks:
      - id: foo
```

This is the renovate config that works for update the refs:

```json
{
  "regexManagers": [
        {
          "fileMatch": ["(^|/)\\.pre-commit-config\\.yaml$"],
          "matchStrings": ["repo: (?<depName>.{2,}?)\n\\s+rev: (?<currentDigest>.*?)\n"],
          "datasourceTemplate": "git-refs",
          "versioningTemplate": "git"
        }
  ]
}

**Comment by benjamb** [comment]
Is there a workaround to make use of `git-tags` as a datasource with `pre-commit`? `github-tags` seems to require an API token to work, which I would prefer to avoid.

Ideally something that doesn't involve using `regexManagers` like the example above, since that relies on keys being in a specific order, which I can't guarantee.
